### PR TITLE
Rewrite v0.3.sh

### PR DESCRIPTION
**Pull request by zrhmn**
If something doesn't make sense, make sure to ask me about it later.

There are 2 notes I want to make:

1. I was able to run bundle update and bundle exec jekyll serve -l
   but the site contains nothing. I don't think that's right, so something
   is missing or wrong in the config, although the script works as intended.
   Fix the config, not the script.

2. This doesn't make sense. This is a lot of work for a problem that does
   not exist. What's different between this, and having an empty boilerplate
   with the complete file and directory structure packed in a zip or tar
   archive? In fact, that's slightly more efficient because the compressed
   pack will probably be smaller than a script which literally contains the
   paths and contents of all the files within it.
   And if the conten is dynmaic (changes based on external factors like
   config options etc.) then the Jekyll template generator itself is the
   best solution.
   Remember, the best solution is that which saves you the most time.
   Automation is cool and everyone should learn to write bash scripts, but
   don't *invent* problems where there are none. The best way to learn
   is also to work on an existing problem, instead of creating a new one
   out of thin air.

I hope that you take these comments in purely academic sense, I don't mean
to put you down, just advising you to avoid an unnecessarily complicated
path. Happy to help either way.

### CODE REVIEW

File: v0.3.sh [hunk 1]
```diff
@@ -1,13 +1,139 @@
...
+  sass_dir: static/styles
+  style: uncompressed
+markdown: kramdown
+EOF
+
+FRONTMATTER_SEP="---"
+
+echo >&2 "setting up index.html"
+cat << EOF > index.html
+$FRONTMATTER_SEP
```
Comment (zrhmn) [inline v0.3.sh:109]: This is here to demonstrate that bash variables can be used inside heredocs and they get replaced by their values.

File: v0.3.sh [hunk 2]
```diff
@@ -1,13 +1,139 @@
...
+title: Home
+laout: frameworl
```
Comment (zrhmn) [inline v0.3.sh:111]: So this is a typo I missed. Probably why I couldn't get Jekyll to build the website. Try fixing this and see if that builds the website properly.

Comment (zrhmn) [reply v0.3.sh:111]: Watch out. There might be more typos too...

My focus was on showing good bash scripting, I don't really know Jekyll that well.

**Comment by xhifi** [comment]
No. You don’t understand. Jekyll’s boilerplate is a default hello world
blog site with preconfigured config and Gemfile. Not only that it also
includes the useless code which I end up eliminating before I start working
on it. And its frustrating as well as time consuming because it's like I
don’t reinvent the wheel but then to use the wheel I have to assemble every
wire and tube and structure of the wheel myself.

I understand the preconfigured default hello world builds in SSGs are  much
more efficient because they usually pull from CDNs where stuff is usually
prewritten. I have no problem with that at all. If it was Gatsby I’d for
sure use gatsby starter because setting up that boilerplate is much more
time consuming and it doesn’t bring in stupid extensions like Rake and/or
in some cases nokogiri and sitemap gems with it. I end up reconfigging
everything. You get my point?

So I wrote a single script, all I do from now on is run this script in a
folder and have the boilerplate of my own personal preference and I know
for sure which gems are doing what and then I add functionality on top of
this boilerplate however I like.

And the worst part about it is that the preconfigured gems and
configuration jekyll hello world starter brings in makes the project way
more bulky on the build side when the project gets bigger. Im saying about
3 seconds wait timers on every build and every livereload build as well.
And no you can’t use incremental builds because they are fiddly and then
half the time you spend doing jekyll clean.

So it’s a pretty nifty solution I found which I can make much much more
sophisticated over time to my liking because still jekyll is a very good
SSG given you integrate it well with a cms such as forestry or contentful.
I'd say good price for the effort.

And the script has to work because I made it work. My only problem again is:
If you see the part of the script where I print <! DOCTYPE html> it’s
wrong. That's why it’s not building with bundle exec jekyll serve or with
livereload or jekyll build ofc. Why? Because there shouldn’t be a space
between exclamation mark and DOCTYPE in the first tag of the main layout.
It gives errors and doesn’t build. The site config and Gemfile are fine.

As of now I can't seem to escape !{string} syntax because bash starts
listening for events EVEN WHEN IN A PRINTF function or echo. Go ahead and
try it. There should be some way to escape it. I tried it with backslash
escape, variable declaration but no.

Any thoughts on that?
tldr; as of now, the only way to make it work is to run the script, go into
the framework.

Also there was an error in 2nd line which I just fixed: there was no &&
between echo and printf so the Gemfile output was garbage.
echo "source \"https://rubygems.org\"" > Gemfile && printf "
\ngem 'jekyll', \'~> 4.2\'\ngem \'webrick\'\ngem \'wdm\', \'>= 0.1.0\'
if Gem.win_platform?\ngem \'eventmachine\', \'~> 1.0.0\'\n
" >> Gemfile

**Comment by xhifi** [comment]
Oh Shiet, You were right about config. But it's sort of a bug. yml files
sometimes have issues with tabs when tabs are coming in from bash or
something like that. I have replaced tabs with spaces and fixed most of the
errors. Also updated the file in repo with a brand new commit. It's a
working script except the <! DOCTYPE html> part ofcourse. Even the bundle
update and jekyll serve as well.

**Comment by zrhmn** [comment]
>No. You don’t understand. Jekyll’s boilerplate is a default hello world
blog site with preconfigured config and Gemfile. Not only that it also
includes the useless code which I end up eliminating before I start working
on it. And its frustrating as well as time consuming because it's like I
don’t reinvent the wheel but then to use the wheel I have to assemble every
wire and tube and structure of the wheel myself.

Sure.

Do this.

1. Generate a project from Jekyll's boilerplate.
2. Delete/change all the files you want.
3. Push it to a GitHub repo _OR_ make a tar.gz of it and keep that somewhere.
4. Now you have your very own boilerplate.
5. Next time you start a project, start with your own boilerplate instead of Jekyll's boilerplate.

What does the script get you that the above method won't?

**Comment by zrhmn** [comment]
> I understand the preconfigured default hello world builds in SSGs are  much
more efficient because they usually pull from CDNs where stuff is usually
prewritten. I have no problem with that at all. If it was Gatsby I’d for
sure use gatsby starter because setting up that boilerplate is much more
time consuming and it doesn’t bring in stupid extensions like Rake and/or
in some cases nokogiri and sitemap gems with it. I end up reconfigging
everything. You get my point?

- SSGs: I don't know what they are and Google doesn't help either.
- I don't know what Gatsby is either (other than a character from an old book).
- Nokogiri? Is that like harakiri?

Look, dude. I don't know all these terms. I told you already I don't know Jekyll. It looks straightforward enough that I could "figure it out" if I needed to, but I'm not concerned with the Jekyll part _at all_. You asked me for help with bash, and I'm telling you I don't understand the purpose of this script. That is all. CDN, no CDN, I don't care.

You definitely seem to know what you're talking about re: Jekyll.  
Just like that, I just know what I'm talking about re: bash and static site builders in general.

**Comment by zrhmn** [comment]
> So I wrote a single script, all I do from now on is run this script in a
folder and have the boilerplate of my own personal preference and I know
for sure which gems are doing what and then I add functionality on top of
this boilerplate however I like.

Let me explain:
Forget about the static site thing for a moment.
I get that you have a script. It seems to do the following:

- Lay out a map of files and directories on disk.
- Push static content (i.e. the content is part of the script, in its entirety, as text) into some of the files.

If the above is not correct then I have misunderstood something fundamental about what you're trying to do with that script. Please let me know if that's the case.

I'm saying if the content is part of the script as static text, how is that different from _actual files on disk_ containing all of content?

So try this:
- Take the same files that you want to keep and the content that you want to keep.
- Zip/tar it up and use that as a starting point.

Right now you run `./v0.3.sh` in an empty directory to set it up. With a tar file you would be doing: `tar xf my_modified_jekyll_boilerplate.tar.xz` instead. That is better than what you're trying to accomplish because you cannot possibly account for all the corner-cases of filesystem management in a bash script. `tar` does all of that for you. You need static content, you got static content. Just unzip/untar it. Why the script?

**Comment by zrhmn** [comment]
> I have replaced tabs with spaces and fixed most of the
errors. Also updated the file in repo with a brand new commit. It's a
working script except the <! DOCTYPE html> part ofcourse. Even the bundle
update and jekyll serve as well.

This is exactly what I'm talking about. Why do you have to do manual white-space management and convert your file into a C-style string in the first place? You don't get proper editor support for CSS/HTML/YAML etc. inside a bash script. Your HTML or YAML, for example, has no syntax highlighting, no autoindent, no syntax checking etc. because it's just a bash string while it's inside the script. What are you gaining by having that as a part of the script instead of as a file on the disk?

Do you see my point?
